1. Prosečno vreme između dva polaska svake od linija

Ucitavam fajl trips.txt koji prikazuje podatke o polascima pojedinacnih linija

In [1]:
import pandas as pd

tripsPath = "../Desktop/asmPodaci/trips.txt"

trips = pd.read_csv(tripsPath)
print(trips)

       route_id        service_id  trip_id    trip_headsign  direction_id  \
0             2     Zimski-Subota  1421378  Spoljašnji Krug             0   
1             2     Zimski-Subota  1421379  Spoljašnji Krug             0   
2             2     Zimski-Subota  1421380  Spoljašnji Krug             0   
3             2     Zimski-Subota  1421381  Spoljašnji Krug             0   
4             2     Zimski-Subota  1421382  Spoljašnji Krug             0   
...         ...               ...      ...              ...           ...   
66048     90004  Zimski-Radni Dan  1400708         Kumodraž             1   
66049     90004  Zimski-Radni Dan  1400709         Kumodraž             1   
66050     90004  Zimski-Radni Dan  1400710         Kumodraž             1   
66051     90004  Zimski-Radni Dan  1400711         Kumodraž             1   
66052     90004  Zimski-Radni Dan  1400712         Kumodraž             1   

       block_id       shape_id  wheelchair_accessible  bikes_allowed  
0   

Izbacujemo nepotrebne kolone iz trips

In [2]:
trips.drop('block_id', inplace=True, axis=1)
trips.drop('shape_id', inplace=True, axis=1)
trips.drop('wheelchair_accessible', inplace=True, axis=1)
trips.drop('bikes_allowed', inplace=True, axis=1)
print(trips)

       route_id        service_id  trip_id    trip_headsign  direction_id
0             2     Zimski-Subota  1421378  Spoljašnji Krug             0
1             2     Zimski-Subota  1421379  Spoljašnji Krug             0
2             2     Zimski-Subota  1421380  Spoljašnji Krug             0
3             2     Zimski-Subota  1421381  Spoljašnji Krug             0
4             2     Zimski-Subota  1421382  Spoljašnji Krug             0
...         ...               ...      ...              ...           ...
66048     90004  Zimski-Radni Dan  1400708         Kumodraž             1
66049     90004  Zimski-Radni Dan  1400709         Kumodraž             1
66050     90004  Zimski-Radni Dan  1400710         Kumodraž             1
66051     90004  Zimski-Radni Dan  1400711         Kumodraž             1
66052     90004  Zimski-Radni Dan  1400712         Kumodraž             1

[66053 rows x 5 columns]


Ucitavam stop_times - podaci o vremenima prolaska pojedinacnih polazaka kroz pojedinacna stajalista

In [3]:
stopTimesPath = "../Desktop/asmPodaci/stop_times.txt"

stop_times = pd.read_csv(stopTimesPath)
print(stop_times)

         trip_id arrival_time departure_time  stop_id  stop_sequence  \
0         665800     20:00:00       20:00:00      638              1   
1         665800     20:01:00       20:01:00      148              2   
2         665800     20:02:00       20:02:00      746              3   
3         665800     20:04:00       20:04:00      845              4   
4         665800     20:06:00       20:06:00      847              5   
...          ...          ...            ...      ...            ...   
1648776  1494173     23:11:00       23:11:00     1289             20   
1648777  1494173     23:13:00       23:13:00     1291             21   
1648778  1494173     23:14:00       23:14:00     1293             22   
1648779  1494173     23:16:00       23:16:00     1295             23   
1648780  1494173     23:17:00       23:17:00     1297             24   

        stop_headsign  pickup_type  drop_off_type  shape_dist_traveled  \
0                 NaN          NaN            NaN            

C:\Users\Andjela\AppData\Local\Temp\ipykernel_21684\1098820719.py:3: DtypeWarning: Columns (5) have mixed types. Specify dtype option on import or set low_memory=False.
  stop_times = pd.read_csv(stopTimesPath)


Uklanjanje nepotrebnih kolona iz stop_times

In [4]:
stop_times.drop('stop_headsign', inplace=True, axis=1)
stop_times.drop('pickup_type', inplace=True, axis=1)
stop_times.drop('drop_off_type', inplace=True, axis=1)
stop_times.drop('shape_dist_traveled', inplace=True, axis=1)
stop_times.drop('timepoint', inplace=True, axis=1)
print(stop_times)

         trip_id arrival_time departure_time  stop_id  stop_sequence
0         665800     20:00:00       20:00:00      638              1
1         665800     20:01:00       20:01:00      148              2
2         665800     20:02:00       20:02:00      746              3
3         665800     20:04:00       20:04:00      845              4
4         665800     20:06:00       20:06:00      847              5
...          ...          ...            ...      ...            ...
1648776  1494173     23:11:00       23:11:00     1289             20
1648777  1494173     23:13:00       23:13:00     1291             21
1648778  1494173     23:14:00       23:14:00     1293             22
1648779  1494173     23:16:00       23:16:00     1295             23
1648780  1494173     23:17:00       23:17:00     1297             24

[1648781 rows x 5 columns]


Spajanje trips i stop_times

In [5]:
tripStopTimes = trips.set_index('trip_id').join(stop_times.set_index('trip_id'))
print(tripStopTimes)

tripStopTimes = tripStopTimes.dropna(subset=['stop_id'])

         route_id        service_id   trip_headsign  direction_id  \
trip_id                                                             
665800        401    Zimski-Nedelja        Pinosava             0   
665800        401    Zimski-Nedelja        Pinosava             0   
665800        401    Zimski-Nedelja        Pinosava             0   
665800        401    Zimski-Nedelja        Pinosava             0   
665800        401    Zimski-Nedelja        Pinosava             0   
...           ...               ...             ...           ...   
1494173       101  Zimski-Radni Dan  Padinska Skela             0   
1494173       101  Zimski-Radni Dan  Padinska Skela             0   
1494173       101  Zimski-Radni Dan  Padinska Skela             0   
1494173       101  Zimski-Radni Dan  Padinska Skela             0   
1494173       101  Zimski-Radni Dan  Padinska Skela             0   

        arrival_time departure_time  stop_id  stop_sequence  
trip_id                                 

In [6]:
tripStopTimes.to_csv("tripStopTimes.txt", sep='\t')

Filtriram sve redove kod kojih je stop_sequence jednako 1 jer to oznacava pocetnu stanicu tj prvi polazak.

In [7]:
tripStopTimes = tripStopTimes[tripStopTimes['stop_sequence'] == 1]
print(tripStopTimes)

         route_id        service_id          trip_headsign  direction_id  \
trip_id                                                                    
665800        401    Zimski-Nedelja               Pinosava             0   
665801         27    Zimski-Nedelja          Trg Republike             1   
665802        401    Zimski-Nedelja  Slavija /Birčaninova/             1   
665803         27    Zimski-Nedelja             Mirijevo 3             0   
665804        401    Zimski-Nedelja               Pinosava             0   
...           ...               ...                    ...           ...   
1494167        31  Zimski-Radni Dan         Studentski Trg             1   
1494168        31  Zimski-Radni Dan              Konjarnik             0   
1494171        31  Zimski-Radni Dan         Studentski Trg             1   
1494172        31  Zimski-Radni Dan              Konjarnik             0   
1494173       101  Zimski-Radni Dan         Padinska Skela             0   

        arr

Provera da li je neki arrival_time razlicit od departure_time za stop_sequence=1; utvrdjeno je da su sve iste pa izbacujem departure_time; Za ostale stop_sequence se razlikuje;
Takodje, provera da li je neki arrival_time null tj da je polje prazno za stop_sequence=1; utvrdjeno je da nema; Za ostale stop_sequence postoje nullovi

In [8]:
if (tripStopTimes['arrival_time'] == tripStopTimes['departure_time']).all():
    print("All values are equal")
else:
    print("All values are not equal")
tripStopTimes.drop('departure_time', inplace=True, axis=1)

tripStopTimes['arrival_time'].isna().any()

All values are equal


False

In [9]:
tripStopTimes.to_csv("tripStopTimesStopSequenceEq1.txt", sep='\t')

Sortiram vrednosti po route_id i arrival_time da bi imali vremena redom da bi bilo lakse za oduzimanje

In [10]:
tripStopTimesSorted = tripStopTimes.sort_values(by=['route_id', 'arrival_time'])
print(tripStopTimesSorted)
tripStopTimesSorted.to_csv("tripStopTimesSort.txt", sep='\t')

         route_id        service_id    trip_headsign  direction_id  \
trip_id                                                              
1421378         2     Zimski-Subota  Spoljašnji Krug             0   
1421406         2     Zimski-Subota  Unutrašnji Krug             1   
1421558         2    Zimski-Nedelja  Spoljašnji Krug             0   
1421586         2    Zimski-Nedelja  Unutrašnji Krug             1   
1423806         2  Zimski-Radni Dan  Spoljašnji Krug             0   
...           ...               ...              ...           ...   
1325337     90004  Zimski-Radni Dan   Studentski Trg             0   
1400709     90004  Zimski-Radni Dan         Kumodraž             1   
1400692     90004  Zimski-Radni Dan         Kumodraž             1   
1400698     90004  Zimski-Radni Dan         Kumodraž             1   
1400697     90004  Zimski-Radni Dan         Kumodraž             1   

        arrival_time  stop_id  stop_sequence  
trip_id                                   

Arrival time je string pa ga pretvaramo u minute i upisujem u novu kolonu: arrival_time_minutes

In [11]:
tripStopTimesSorted['arrival_time'] = pd.to_timedelta(tripStopTimesSorted['arrival_time'])
tripStopTimesSorted['arrival_time_minutes'] = tripStopTimesSorted['arrival_time'].dt.total_seconds() / 60
print(tripStopTimesSorted)

         route_id        service_id    trip_headsign  direction_id  \
trip_id                                                              
1421378         2     Zimski-Subota  Spoljašnji Krug             0   
1421406         2     Zimski-Subota  Unutrašnji Krug             1   
1421558         2    Zimski-Nedelja  Spoljašnji Krug             0   
1421586         2    Zimski-Nedelja  Unutrašnji Krug             1   
1423806         2  Zimski-Radni Dan  Spoljašnji Krug             0   
...           ...               ...              ...           ...   
1325337     90004  Zimski-Radni Dan   Studentski Trg             0   
1400709     90004  Zimski-Radni Dan         Kumodraž             1   
1400692     90004  Zimski-Radni Dan         Kumodraž             1   
1400698     90004  Zimski-Radni Dan         Kumodraž             1   
1400697     90004  Zimski-Radni Dan         Kumodraž             1   

           arrival_time  stop_id  stop_sequence  arrival_time_minutes  
trip_id          

Racunam razliku izmedju vremena i upisujem u novu kolonu - diff. Grupisem po route_id jer su  bitni polasci za pojedinacnu stanicu. Takodje, mora se grupisati po service_id da ne bi racunao razliku vremena polaska iz npr petka i subote. Takodje, grupisanje po direction_id da ne bi racunao razliku polaska vremena izmedju pocetne i krajnje stanice.

In [12]:
tripStopTimesSorted['diff'] = tripStopTimesSorted.groupby(['route_id', 'service_id', 'direction_id'])['arrival_time_minutes'].diff().fillna(0)
tripStopTimesSorted.to_csv("tripStopTimesGroupBy.txt", sep='\t')
print(tripStopTimesSorted)

         route_id        service_id    trip_headsign  direction_id  \
trip_id                                                              
1421378         2     Zimski-Subota  Spoljašnji Krug             0   
1421406         2     Zimski-Subota  Unutrašnji Krug             1   
1421558         2    Zimski-Nedelja  Spoljašnji Krug             0   
1421586         2    Zimski-Nedelja  Unutrašnji Krug             1   
1423806         2  Zimski-Radni Dan  Spoljašnji Krug             0   
...           ...               ...              ...           ...   
1325337     90004  Zimski-Radni Dan   Studentski Trg             0   
1400709     90004  Zimski-Radni Dan         Kumodraž             1   
1400692     90004  Zimski-Radni Dan         Kumodraž             1   
1400698     90004  Zimski-Radni Dan         Kumodraž             1   
1400697     90004  Zimski-Radni Dan         Kumodraž             1   

           arrival_time  stop_id  stop_sequence  arrival_time_minutes  diff  
trip_id    

Izbaciti vrste gde je diff tj. razlika u vremenima jednaka nuli(oznacava prvi polazak u tom danu)

In [13]:
print(tripStopTimesSorted.shape[0])
tripStopTimesWithoutZeros = tripStopTimesSorted[tripStopTimesSorted['diff']!=0]
print(tripStopTimesWithoutZeros.shape[0])

66029
64818


Racunanje proseka kolone diff da bi nasla prosecno vreme polaska buseva sa stanica.
Spajanje sa routes da bi dobila puno ime linije i uklanjanje nepotrebnih kolona

In [14]:
averageDepartureTime = tripStopTimesWithoutZeros.groupby(['route_id', 'direction_id','trip_headsign'], as_index= False)['diff'].mean()

averageDepartureTime.rename(columns={"diff": "mean"}, inplace=True)

routesPath = "../Desktop/asmPodaci/routes.txt"
routes = pd.read_csv(routesPath)
routes.drop('agency_id', inplace=True, axis=1)
routes.drop('route_desc', inplace=True, axis=1)
routes.drop('route_type', inplace=True, axis=1)
routes.drop('route_url', inplace=True, axis=1)
routes.drop('route_color', inplace=True, axis=1)
routes.drop('route_text_color', inplace=True, axis=1)
            
averageDepartureTime = averageDepartureTime.set_index('route_id').join(routes.set_index('route_id'))
print(averageDepartureTime)

          direction_id            trip_headsign       mean route_short_name  \
route_id                                                                      
2                    0          Spoljašnji Krug  13.258427                2   
2                    1          Unutrašnji Krug  13.258427                2   
5                    0                Ustanička  12.834646                5   
5                    1  Kalemegdan /Donji Grad/  12.859922                5   
6                    0                Ustanička  24.457364                6   
...                ...                      ...        ...              ...   
90002                1             Petlovo Brdo  15.205128               E2   
90003                0   Novi Beograd /Blok 45/  20.000000               E6   
90003                1               Mirijevo 4  20.000000               E6   
90004                0           Studentski Trg  12.000000               E9   
90004                1                 Kumodraž  11.

In [15]:
averageDepartureTime.to_csv("tripStopTimesMean.txt", sep='\t')
averageDepartureTime = averageDepartureTime.sort_values(by=['mean'])
print(averageDepartureTime)
averageDepartureTime.to_csv("tripStopTimesMeanSorted.txt", sep='\t')

          direction_id             trip_headsign        mean route_short_name  \
route_id                                                                        
20022                0               Trg Slavija    5.190616              22L   
20022                1            Studentski Trg    5.212610              22L   
41                   0                 Banjica 2    6.685714               41   
41                   1            Studentski Trg    6.776699               41   
26                   1         Dorćol /Dunavska/    6.782869               26   
...                ...                       ...         ...              ...   
50603                0                 Ugrinovci   95.000000             603N   
50401                1          Dorćol/Dunavska/  100.000000             401N   
50401                0                  Pinosava  100.000000             401N   
607                  1  Novi Beograd /Paviljoni/  105.000000              607   
607                  0      

Gledam ko ima najveci broj polazaka u proseku a ko najmanji

In [16]:
min = averageDepartureTime[averageDepartureTime['mean'] == averageDepartureTime['mean'].min()]
print(min)

          direction_id trip_headsign      mean route_short_name  \
route_id                                                          
20022                0   Trg Slavija  5.190616              22L   

                       route_long_name  
route_id                                
20022     Studentski Trg - Trg Slavija  


In [17]:
max = averageDepartureTime[averageDepartureTime['mean'] == averageDepartureTime['mean'].max()]
print(max)

          direction_id trip_headsign   mean route_short_name  \
route_id                                                       
607                  0        Surčin  105.0              607   

                            route_long_name  
route_id                                     
607       Novi Beograd /Paviljoni/ - Surčin  


2. Linije sa najvećim brojem stanica i stanice sa najvećim brojem linija

2.0. Za pronalazak linija sa najvecim brojem stanica uzimam tabele trips i stop_times koje su spojene po trip_id.
Grupisem po broju linije da ne bi brojao duplo. Biram max po stop_sequence jer on broji stanice po liniji.
Ima 1214 redova: u postavci pise da ima 214 lnija(sto je i provereno u kodu sa unique); kod mene se dobije 208.
208*6(3 service_id * 2 direction_id) = 1248, medjutim neke stanice nemaju service_id za nedelju pa ima manje - ipak nema ovog problema kada se uradi group, vec kada se izbace neke vrste(npr. Zimski-Radni Dan)

In [18]:
tripStopTimes = trips.set_index('trip_id').join(stop_times.set_index('trip_id'))
print(tripStopTimes.shape[0])

tripStopTimes = tripStopTimes.dropna(subset=['stop_id'])

print(tripStopTimes.shape[0])

uniqueRouteId = tripStopTimes["route_id"].unique()
print(uniqueRouteId.shape[0])

maxStopSequence = tripStopTimes.groupby(['route_id','trip_headsign'], as_index= False)['stop_sequence'].max()
maxStopSequence.rename(columns={"stop_sequence": "max"}, inplace=True)
print(maxStopSequence)

maxBrojStanica = maxStopSequence[maxStopSequence['max'] == maxStopSequence['max'].max()]
print(maxBrojStanica)

maxStopSequence = tripStopTimes.groupby(['route_id'], as_index= False)['stop_sequence'].max()
maxStopSequence.rename(columns={"stop_sequence": "max"}, inplace=True)
print(maxStopSequence)

maxBrojStanica = maxStopSequence[maxStopSequence['max'] == maxStopSequence['max'].max()]
print(maxBrojStanica)

maxStopSequence.to_csv("maxBrojStanica.txt", sep='\t')

sortedLinesByStops = maxStopSequence.sort_values(by=['max'], ascending=False);
sortedLinesByStops.to_csv("maxBrojStanicaSortirano.txt", sep='\t')
print(sortedLinesByStops)

1648805
1648781
214
     route_id            trip_headsign   max
0           2          Spoljašnji Krug  17.0
1           2          Unutrašnji Krug  16.0
2           5  Kalemegdan /Donji Grad/  22.0
3           5                Ustanička  21.0
4           6                Tašmajdan  13.0
..        ...                      ...   ...
411     90002            Trg Republike  33.0
412     90003               Mirijevo 4  41.0
413     90003   Novi Beograd /Blok 45/  44.0
414     90004                 Kumodraž  26.0
415     90004           Studentski Trg  27.0

[416 rows x 3 columns]
     route_id trip_headsign   max
381     50601     Dobanovci  88.0
     route_id   max
0           2  17.0
1           5  22.0
2           6  13.0
3           7  32.0
4           9  34.0
..        ...   ...
209     53047  22.0
210     90001  33.0
211     90002  33.0
212     90003  44.0
213     90004  27.0

[214 rows x 2 columns]
     route_id   max
194     50601  88.0
     route_id   max
194     50601  88.0
190 

Linija sa najvećim brojem stanica je 50601, odnosno 601N - Dobanovci

2.1. Racunanje broja linija po stanici

Spajam trip i stop_times i izbacujem arrival_time jer mi za ovo izracunavanje nije bitno. Sada izbacujem duplikate vezane za stanicu(stop_id) i liniju(route_id) jer ima vise polazaka linija u toku dana. Onda grupisem po stanici i prebrojim linije.

In [19]:
tripStopTimes = trips.set_index('trip_id').join(stop_times.set_index('trip_id'))
print(tripStopTimes.shape[0])

tripStopTimes = tripStopTimes.dropna(subset=['stop_id'])

tripStopTimes.drop('arrival_time', inplace=True, axis=1)

tripStopTimesUnique = tripStopTimes.drop_duplicates(subset=['route_id','stop_id'])
print(tripStopTimesUnique)

numberOfLinesByStops = tripStopTimesUnique.groupby(['stop_id'], as_index= False).count()
print(numberOfLinesByStops)

maxLinesByStop = numberOfLinesByStops.sort_values(by=['stop_sequence'], ascending=False);
print(maxLinesByStop)


1648805
         route_id        service_id trip_headsign  direction_id  \
trip_id                                                           
665800        401    Zimski-Nedelja      Pinosava             0   
665800        401    Zimski-Nedelja      Pinosava             0   
665800        401    Zimski-Nedelja      Pinosava             0   
665800        401    Zimski-Nedelja      Pinosava             0   
665800        401    Zimski-Nedelja      Pinosava             0   
...           ...               ...           ...           ...   
1470996     20022  Zimski-Radni Dan   Trg Slavija             0   
1470996     20022  Zimski-Radni Dan   Trg Slavija             0   
1470996     20022  Zimski-Radni Dan   Trg Slavija             0   
1470996     20022  Zimski-Radni Dan   Trg Slavija             0   
1470996     20022  Zimski-Radni Dan   Trg Slavija             0   

        departure_time  stop_id  stop_sequence  
trip_id                                         
665800        20:00:00

Spajam tabelu tripsStopTimes i stops da bih videla naziv stanice sa najvecim brojem linija.
Dobijeno je da najveci broj linija ima stanica 298 - Brankov Most sa 24 linija

In [20]:
stopsPath = "../Desktop/asmPodaci/stops.txt"

stops = pd.read_csv(stopsPath)

stops.drop('stop_code', inplace=True, axis=1)
stops.drop('stop_desc', inplace=True, axis=1)
stops.drop('stop_lat', inplace=True, axis=1)
stops.drop('stop_lon', inplace=True, axis=1)
stops.drop('zone_id', inplace=True, axis=1)
stops.drop('stop_url', inplace=True, axis=1)
stops.drop('location_type', inplace=True, axis=1)
stops.drop('parent_station', inplace=True, axis=1)
stops.drop('stop_timezone', inplace=True, axis=1)
stops.drop('wheelchair_boarding', inplace=True, axis=1)

stopsStopTimes = stops.set_index('stop_id').join(maxLinesByStop.set_index('stop_id'))
stopsStopTimes = stopsStopTimes.sort_values(by=['stop_sequence'], ascending=False);
print(stopsStopTimes)
stopsStopTimes.to_csv("maxBrojLinijaSortirano.txt", sep='\t')

                       stop_name  route_id  service_id  trip_headsign  \
stop_id                                                                 
298.0               Brankov Most        24          24             24   
149.0            Karađorđev Park        23          23             23   
496.0                      Sajam        23          23             23   
503.0              Ada Ciganlija        23          23             23   
148.0            Karađorđev Park        22          22             22   
...                          ...       ...         ...            ...   
2410.0   Kumodraž (Stepina Kuća)         1           1              1   
2412.0           Ripanj /Centar/         1           1              1   
2413.0           Ripanj /Centar/         1           1              1   
1512.0                 Zaklopača         1           1              1   
7970.0                  Manastir         1           1              1   

         direction_id  departure_time  stop_sequen

3. Ocekivani dnevni promet stanice kao ukupan ocekivani broj stajanja linije na toj stanici

In [21]:
tripStopTimes = trips.set_index('trip_id').join(stop_times.set_index('trip_id'))

tripStopTimes = tripStopTimes.dropna(subset=['stop_id'])

tripStopTimes.drop('arrival_time', inplace=True, axis=1)
tripStopTimes.drop('service_id', inplace=True, axis=1)
tripStopTimes.drop('direction_id', inplace=True, axis=1)
tripStopTimes.drop('stop_sequence', inplace=True, axis=1)
print(tripStopTimes)

dailyCountLinesByStop = tripStopTimes.groupby(['stop_id'], as_index= False).count()
print(dailyCountLinesByStop)

stopsWithMaxDailyCount = dailyCountLinesByStop.sort_values(by=['route_id'], ascending=False);
print(stopsWithMaxDailyCount)

         route_id   trip_headsign departure_time  stop_id
trip_id                                                  
665800        401        Pinosava       20:00:00    638.0
665800        401        Pinosava       20:01:00    148.0
665800        401        Pinosava       20:02:00    746.0
665800        401        Pinosava       20:04:00    845.0
665800        401        Pinosava       20:06:00    847.0
...           ...             ...            ...      ...
1494173       101  Padinska Skela       23:11:00   1289.0
1494173       101  Padinska Skela       23:13:00   1291.0
1494173       101  Padinska Skela       23:14:00   1293.0
1494173       101  Padinska Skela       23:16:00   1295.0
1494173       101  Padinska Skela       23:17:00   1297.0

[1648781 rows x 4 columns]
      stop_id  route_id  trip_headsign  departure_time
0         1.0       616            616             616
1         2.0       616            616             616
2         3.0       616            616             61

In [22]:
stopsDaily = stops.set_index('stop_id').join(stopsWithMaxDailyCount.set_index('stop_id'))
stopsDaily = stopsDaily.sort_values(by=['route_id'], ascending=False);
print(stopsDaily)
stopsDaily.to_csv("maxDnevniPrometStanice.txt", sep='\t')

                   stop_name  route_id  trip_headsign  departure_time
stop_id                                                              
298.0           Brankov Most      4063           4063            4063
496.0                  Sajam      4051           4051            4051
495.0                  Sajam      3999           3999            3999
503.0          Ada Ciganlija      3985           3985            3985
149.0        Karađorđev Park      3916           3916            3916
...                      ...       ...            ...             ...
2847.0   Novi Beograd (Ušće)        10             10              10
3735.0       Novi Železnik 1         9              9               9
1243.0      Mihajla Šolohova         6              6               6
1245.0         Milana Ajvaza         6              6               6
1247.0     Valjevskog Odreda         6              6               6

[3130 rows x 4 columns]


4. Svakoj stanici dodeliti tip na osnovu vrsta prevoza koje prolaze njom (stanica autobusa, tramvaja, 
trolejbusa ili meštovita stanica). Koliko ima stanica svakog tipa?

In [23]:
tripStopTimes = trips.set_index('trip_id').join(stop_times.set_index('trip_id'))

tripStopTimes = tripStopTimes.dropna(subset=['stop_id'])

tripStopTimes.drop('arrival_time', inplace=True, axis=1)
tripStopTimes.drop('service_id', inplace=True, axis=1)
tripStopTimes.drop('direction_id', inplace=True, axis=1)
tripStopTimes.drop('stop_sequence', inplace=True, axis=1)

tripStopTimes =  tripStopTimes.drop_duplicates(subset=['route_id','stop_id'])

routesPath = "../Desktop/asmPodaci/routes.txt"
routes = pd.read_csv(routesPath)
routes.drop('agency_id', inplace=True, axis=1)
routes.drop('route_desc', inplace=True, axis=1)
routes.drop('route_url', inplace=True, axis=1)
routes.drop('route_color', inplace=True, axis=1)
routes.drop('route_text_color', inplace=True, axis=1)

tripStopTimesRoutes = tripStopTimes.set_index('route_id').join(routes.set_index('route_id'))
tripStopTimesRoutes['route_type'] = tripStopTimesRoutes['route_type'].replace([702], 3)
print(tripStopTimesRoutes)
tripStopTimesRoutes.to_csv("tripStopTimesRoutes.txt", sep='\t')



            trip_headsign departure_time  stop_id route_short_name  \
route_id                                                             
2         Spoljašnji Krug       04:10:00      7.0                2   
2         Spoljašnji Krug       04:14:00      5.0                2   
2         Spoljašnji Krug       04:18:00     71.0                2   
2         Spoljašnji Krug       04:20:00     73.0                2   
2         Spoljašnji Krug       04:22:00     52.0                2   
...                   ...            ...      ...              ...   
90004            Kumodraž       07:47:00    688.0               E9   
90004            Kumodraž       07:48:00    690.0               E9   
90004            Kumodraž       07:49:00    692.0               E9   
90004            Kumodraž       07:49:00    694.0               E9   
90004            Kumodraž       07:50:00   2199.0               E9   

                            route_long_name  route_type  
route_id                       

In [24]:
tripStopTimesRoutes = tripStopTimesRoutes.set_index('stop_id').join(stops.set_index('stop_id'))
tripStopTimesRoutes = tripStopTimesRoutes.groupby(['stop_id', 'stop_name'])['route_type'].value_counts().unstack().fillna(0)

conditionBus = tripStopTimesRoutes[3]>0
conditionTramvaj = tripStopTimesRoutes[0]>0
conditionTrola = tripStopTimesRoutes[800]>0
conditionMix = ((tripStopTimesRoutes[3]>0) & (tripStopTimesRoutes[0]>0)) | ((tripStopTimesRoutes[3]>0) & (tripStopTimesRoutes[800]>0)) | ((tripStopTimesRoutes[0]>0) & (tripStopTimesRoutes[800]>0))
#tripStopTimesRoutes = tripStopTimesRoutes[condition]

tripStopTimesRoutes['type'] = 'mesovita'
tripStopTimesRoutes.loc[conditionBus, "type"] = "autobus"
tripStopTimesRoutes.loc[conditionTramvaj, "type"] = "tramvaj"
tripStopTimesRoutes.loc[conditionTrola, "type"] = "trola"
tripStopTimesRoutes.loc[conditionMix, "type"] = "mix"
print(tripStopTimesRoutes)

tripStopTimesRoutes.to_csv("tripStopTimesRoutesTypes.txt", sep='\t')

numberOfBuses = tripStopTimesRoutes.groupby(['type'], as_index= False).count()
print(numberOfBuses)

route_type                            0    3  800     type
stop_id stop_name                                         
1.0     Kalemegdan (Donji Grad)     3.0  0.0  0.0  tramvaj
2.0     Tadeuša Košćuška            3.0  0.0  0.0  tramvaj
3.0     Tadeuša Košćuška            3.0  0.0  0.0  tramvaj
4.0     Kalemegdan                  2.0  0.0  0.0  tramvaj
5.0     Kalemegdan                  2.0  0.0  0.0  tramvaj
...                                 ...  ...  ...      ...
7941.0  Simić - R                   0.0  1.0  0.0  autobus
7942.0  Simić - R                   0.0  1.0  0.0  autobus
7968.0  Stara Lipovica - Okretnica  0.0  1.0  0.0  autobus
7969.0  Manastir                    0.0  1.0  0.0  autobus
7970.0  Manastir                    0.0  1.0  0.0  autobus

[3130 rows x 4 columns]
route_type     type     0     3   800
0           autobus  2901  2901  2901
1               mix   151   151   151
2           tramvaj    74    74    74
3             trola     4     4     4
